core/src/server/mod.rs

```diff
@@ -32,3 +32,48 @@ pub mod helpers;
 pub mod host_filtering;
 /// JSON-RPC "modules" group sets of methods that belong together and handles method/subscription registration.
 pub mod rpc_module;
+
+use self::rpc_module::DisconnectError;
+use crate::error::SubscriptionAcceptRejectError;
+use jsonrpsee_types::error::CallError;
+use rpc_module::SubscriptionMessage;
+
+type SubscriptionCloseResponse = Option<SubscriptionMessage>;
+
+/// Extension trait that converts errors into an optional error message.
+///
+/// Internally it converts an `Result<T, Error>` into `Result<T, Option<Error>`
+/// which forces users to convert `Error` to `Option<SubscriptionMessage>`
+/// where `Some(msg)` represents that the message is sent as subscription notification error
+/// and `None` doesn't do anything.
+///
+/// This is implemented for the types in jsonrpsee where the behavior is not application dependent.
+/// For other types you have to implement this trait or deal with it manually.
+pub trait MapSubscriptionError<T> {
```

Comment:
I'm not sure that we should have this trait, it works good for user-defined types but for other types from jsonrpsee such as `TrySendError`, `SendTimeoutError` then folks has to move them into a new type to not bump orphan rule issues.

Reply:
also to `map_err` works now so perhaps better just to have that

Reply:
Mmm I'm not super keen on this trait; if I was going for any trait it'd be something like:

```rust
trait IntoResponse {
    fn into_response(self) -> SubscriptionCloseResponse
}

impl IntoResponse for Option<T> where T: Into<SubscriptionMessage> {
    fn into_response(self) ->  SubscriptionCloseResponse {
         // None means nothing returned, Some(msg) means a non-error msg is returned
    }
}

impl IntoResponse for Result<T, E> {
    fn into_response(self) ->  SubscriptionCloseResponse {
          // Err(e) means error returned, Ok(m) means a non-error msg is returned
    }
}
```

But should `Ok(())` and `None` mean that no message is returned at all, or with `Ok(())` would you get back `[]` as a message? It feels like it would be safer if no message came back (otherwise using `Result` there is no way _not_ to return some message if all goes ok), and you just have to explicitly send one yourself (or register one to be sent on the sink).

So I guess I'd only support `Result<(),IntoError>` and `Option<()>`, and then we can ergonomically return errors using `Result` and `?` or ergonomically return but emit no errors with `Option` and `None`.

So one could write:

```rust
module.register_subscription("my_sub", "my_sub", "my_unsub", |_, pending, _| async move {
	// Do stuff; can use ? to return errors here
    pending.accept().await?;

	Ok(())
})
```

Or one can opt to not return errors like this:

```rust
module.register_subscription("my_sub", "my_sub", "my_unsub", |_, pending, _| async move {
	// Due to return type, need to `.ok()` explicitly to throw away errors.
    pending.accept().await.ok()?;

	None
})
```

And maybe if you want to send a default closing message some api like:

```rust
module.register_subscription("my_sub", "my_sub", "my_unsub", |_, pending, _| async move {
    let mut sink = pending.accept().await.ok()?;

    // register message to be sent when closes (presumably if an error is returned because we
    // allow this via `Result`, we will either send this just prior to the error, or not send it at all 
    // (don't know that it matters which one!):
    sink.send_on_close(|| SomeMessage);

	None
})
```

But if all of that is sortof annoying or too much effort, then the alternative which I also think is perfectly reasonable is to only return `()` (and force error handling to be done by the user), or only return `Result<(), SomeError>` to force the user to do `pending.accept().await? else { /* handle error manually each time */ }` but make it easy if the user wants to return errors by default.

Really, the only thing I don't like about the currently released interface is that you can return the error and then it's silently discarded, which I think may be surprising.

Reply:
I guess the current API you've settled on without the trait is pretty much equivalent to returning `()` in terms of usability, except for allowing the user to return some message (or error I think?) if they like. So I'm also happy with that offhand if that's what you prefer!

Reply:
The reason why I didn't go down the `into_response` route is because it didn't work with `?` but maybe I did something wrong when I tried it.

Reply:
so yeah, I got it working as well but as the type inference forces the `async block` to return `Result or Option` with the design which makes it not possible to ignore some error but in another scenario return an error if that makes sense :)

[example here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d7429031e971510c4ffd00f347df433b)

Reply:
> which makes it not possible to ignore some error but in another scenario return an error in the same async block

I think you mean that you want some errors to be ignored (eg connection closed) when `?` is applied on them, while letting other errors propagate to the user?

`IntoResponse` could always expect an error enum to be returned and then "connection closed" style errors can be manually filtered out if they are seen? Or errors can impl `Into<ResponseError>` which would have the error info and also note whether it should be sent to the user or not (http libraries sometimes do similar stuff; sending 4xx errors to users and only logging 5xx errors on the server while sending the user some generic message)

But I might be missing the point so apologies if so!

Reply:
Yeah, that's true maybe I'm a bit narrow minded but initially the idea was that `return Option == ignore error` and `return Err == always propagate the error` to make it explicit for the user.

Having what you suggested is nice but isn't really what I had in mind and easy to "bail early with ?`....

Reply:
> but initially the idea was that return Option == ignore error and return Err == always propagate the error to make it explicit for the user.

Yup that was definitely the thinking but I thought you were raising that not all errors should be returned by default, but I may have misunderstood!

Reply:
yeah sorry I was more picking hair on the that you can't really return `Result and Option` from the substration callback and tried to elaborate that it's ok but yeah we could ignore some error but it doesn't really matter if the connection is closed anyway :P

I think I prefer your suggestion anyway, I'll give it a try again